sql/index.go

```diff
@@ -120,24 +120,36 @@ type IndexLookup struct {
 
 var emptyLookup = IndexLookup{}
 
-func (il IndexLookup) IsEmpty() bool {
+func (il *IndexLookup) IsEmpty() bool {
 	return il.Index == nil
 }
 
-func (il IndexLookup) String() string {
+func (il *IndexLookup) String() string {
 	pr := NewTreePrinter()
 	_ = pr.WriteNode("IndexLookup")
 	pr.WriteChildren(fmt.Sprintf("index: %s", il.Index), fmt.Sprintf("ranges: %s", il.Ranges.String()))
 	return pr.String()
 }
 
-func (il IndexLookup) DebugString() string {
+func (il *IndexLookup) DebugString() string {
 	pr := NewTreePrinter()
 	_ = pr.WriteNode("IndexLookup")
 	pr.WriteChildren(fmt.Sprintf("index: %s", il.Index), fmt.Sprintf("ranges: %s", il.Ranges.DebugString()))
 	return pr.String()
 }
 
+// Reverse flips the order of the ranges in the lookup, and marks the lookup as reversed.
+// This will do nothing if the lookup is already reversed
+func (il *IndexLookup) Reverse() {
```

Comment:
It feels odd that the optimization has to both set `IsReverse` and then also call this method.

I think a simpler solution would be to change the iteration order in `IndexedTable::PartitionRows` if `IsReverse` is true. That way, it's no longer possible to accidentally generate an "invalid" IndexedTableAccess where the ordering of the filters no longer matches the value of `IsReversed`.

Reply:
Having the range order be in the debug plan might be useful. An alternative might be avoiding an in-place modify, make a `NewLookup` with a reverse argument and sort on construction. I agree the in between is weird.

Reply:
The optimization doesn't have to set `IsReverse` and call this method; this method sets `IsReverse`. I did it this way to avoid accidentally generating Lookups that were marked as reversed, but had filters in the wrong order. 

There are a lot of things that implement `IndexedTable` so I'd have to add a check for all of them. Unless I'm misunderstanding I don't think that's easier

Reply:
@nicktobey would you be more comfortable with a private attribute?

Reply:
My mistake; I misread the diff.

> There are a lot of things that implement IndexedTable so I'd have to add a check for all of them. Unless I'm misunderstanding I don't think that's easier

IndexedTable, not IndexableTable. Two types with confusingly similar names. Currently IndexedTable is the only place that reads `IsReversed`.

Overall, Max raises a good point about avoiding the in-place modification. We usually try for nodes to be immutable. He also makes a good point about the range order being in the debug plan for debugging purposes.

I agree with the idea of a `NewLookup` method that constructs a new node.

Maybe IndexedTable::PartitionRows could have an assert that the filters are in the expected order, and panic if they're aren't? Just as a sanity check.